chainstate/tx-verifier/src/transaction_verifier/mod.rs

```diff
@@ -326,27 +368,89 @@ where
                     Some(res)
                 }
+                TxOutput::CreateDelegationId(spend_destination, target_pool) => {
+                    let res = self
+                        .accounting_delta_adapter
+                        .operations(tx_source)
+                        .create_delegation_id(
+                            *target_pool,
+                            spend_destination.clone(),
+                            input0.outpoint(),
+                        )
+                        .map(|(_, undo)| undo)
+                        .map_err(ConnectTransactionError::PoSAccountingError);
+                    Some(res)
+                }
+                TxOutput::DelegateStaking(amount, delegation_id) => {
+                    let res = self
+                        .accounting_delta_adapter
+                        .operations(tx_source)
+                        .delegate_staking(*delegation_id, *amount)
+                        .map_err(ConnectTransactionError::PoSAccountingError);
+                    Some(res)
+                }
+                TxOutput::SpendShareFromDelegation(_, _, _, _) => {
+                    // TODO: check for delegation_id mismatch
+                    None
+                }
                 TxOutput::Transfer(_, _)
                 | TxOutput::LockThenTransfer(_, _, _)
                 | TxOutput::Burn(_)
                 | TxOutput::ProduceBlockFromStake(_, _) => None,
             })
             .collect::<Result<Vec<_>, _>>()?;
 
-        if !tx_undo.is_empty() {
+        // delete delegation if the balance is 0 and the pool has been decommissioned
+        let delete_delegation_undo = match check_for_delegation_cleanup {
+            Some(delegation_id) => {
+                let accounting_view = self.accounting_delta_adapter.accounting_delta();
+                let delegation_balance =
+                    accounting_view.get_delegation_balance(delegation_id)?.ok_or(
+                        ConnectTransactionError::DelegationDataNotFound(delegation_id),
+                    )?;
+                if delegation_balance == Amount::ZERO {
+                    let delegation_data =
+                        accounting_view.get_delegation_data(delegation_id)?.ok_or(
+                            ConnectTransactionError::DelegationDataNotFound(delegation_id),
+                        )?;
+                    if !accounting_view.pool_exists(*delegation_data.source_pool())? {
+                        Some(
+                            self.accounting_delta_adapter
+                                .operations(tx_source)
+                                .delete_delegation_id(delegation_id)?,
+                        )
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            }
+            None => None,
+        };
```

Comment:
I am wondering if the cascade of conditionals here can be simplified somewhat with combinators like `map`, `filter`, `flat_map` and such. The issue I see is the error handling which would force the use of `transpose` and similar making it potentially less readable.

Reply:
I tried to play with `map` and `bool::then` but didn't like it. You are right that error handling here becomes a mess